teps/0033-tekton-feature-gates.md

```diff
@@ -0,0 +1,153 @@
+---
+status: proposed
+title: Tekton Feature Gates
+creation-date: '2020-11-20'
+last-updated: '2020-11-20'
+authors:
+- '@vdemeester'
+- '@frittoli'
+---
+
+# TEP-0033: Tekton Feature Gates
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Requirements](#requirements)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Several Tekton components, such as [pipeline](https://github.com/tektoncd/pipeline)
+and [triggers](https://github.com/tektoncd/triggers), expose features to their
+users through APIs. Today APIs can have a different level of maturity, and
+accordingly of stability, which can range between *alpha*, *beta* and *stable*
+(also known as GA, general availability).
+
+An API reaches a higher level of maturity through validation as well as through
+the assessment of the feedback provided by our users. Marking an API as *beta*, for
+instance, means that we are happy to put backward incompatible changes to the
+API behind a deprecation window of 9 months.
+
+The stability of the API is important to foster adoption and a growing ecosystem of
+users, products and services on top of Tekton. At the same time, though, it risks
+to inhibit innovation and development of new features. 
+
+Aim of this TEP is to define requirements, processes, tools and best practises to
+support the introduction of changes and new features in Tekton, while preserving
+the stability required by our users.
+
+
+## Motivation
+
+
+The Tekton community wants to provide users with a stable API they can rely
+upon. At the same time they want to introduce new features to meet the needs of
+its thriving community of users.
+
+In the pipeline project we defined and API compatibility policy to provide our
+users with the API stability they require. The same policy, however, risks to
+inhibit and slow down development of new features, especially when they require
+changes to the API, since any change made will have to be supported for at
+least 9 months.
+
+The tools we have today are deprecation notices in the release notes, a
```

Comment:
Another tool is using annotations to control a feature on a per resource (Task/Template etc.) basis. 

e.g. in triggers we have a custom annotaiton for templates: https://github.com/tektoncd/triggers/blob/master/docs/triggertemplates.md#escaping-quoted-strings

and we proposed a custom annotation for https://github.com/tektoncd/pipeline/issues/3257

Reply:
Right, I think the annotation/label can be used "in addition to" a way to enable/disable globally.